dart/common/MutexReference.hpp

```diff
@@ -0,0 +1,118 @@
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_MUTEXREFERENCE_HPP_
+#define DART_COMMON_MUTEXREFERENCE_HPP_
+
+#include <mutex>
+#include <set>
+
+namespace dart {
+namespace common {
+
+/// Mutex is a wrapper class of single or multiple std::mutex to provide unified
+/// interface that guarantees deadlock-free locking and unlocking of the
+/// internal mutex(es).
+///
+/// This class is compatible to BasicLockable concept so that it can be used
+/// as a template parameter that requires this concept such as std::lock_guard.
+class MutexReference
+{
+public:
+  /// Default construtor
+  constexpr MutexReference() noexcept = default;
+
+  /// Default destructor
+  virtual ~MutexReference() = default;
+
+  /// Locks mutex that this class references; blocks if one of the mutexes are
+  /// not avaliable.
+  virtual void lock() = 0;
+
+  /// Tries to lock the mutexes that this class references; returns false if
+  /// one of the mutexes is not avaliable.
+  virtual bool try_lock() noexcept = 0;
+
+  /// Unlocks the mutexes.
+  virtual void unlock() noexcept = 0;
+
+protected:
+  /// Copy construction is not allowed.
+  MutexReference(const MutexReference&) = delete;
+};
+
+/// This class references a single mutex.
+class SingleMutexReference final : public MutexReference
+{
+public:
+  /// Constructor from a single mutex.
+  explicit SingleMutexReference(std::mutex& mutex) noexcept;
+
+  // Documentation inherited
+  void lock() override;
+
+  // Documentation inherited
+  bool try_lock() noexcept override;
+
+  // Documentation inherited
+  void unlock() noexcept override;
+
+private:
+  /// Mutex this class references.
+  std::mutex& mMutex;
+};
+
+/// This class references multiple mutexes.
+class MultiMutexReference final : public MutexReference
```

Comment:
Just a thought: Template this class on an iterator type. Then, we could replace the `std::set<std::mutex*>&` with a pair of iterators. That would make this class support any type of collection that: (a) has a forward iterator and (b) supports multiple iteration.

This would be a big step towards @mxgrey's request to hide this type of implementation detail behind the API.

Reply:
Done.

I templatize this class for a lockable type instead of an iterator. Instead, the constructor is templatized for an iterator and check the compatibility between the lockable type of this class and the value type of iterator using `static_assert`.

This is because (1) it would allow us to do type aliasing for the lockable types as
```cpp
using SingleMutexReference = SingleLockableReference<std::mutex>;
using MultiMutexReference = MultiLockableReference<std::mutex>;
```
and (2) it would be authentically better (I think) when we create `MultiLocableReference`. For example,
```cpp
  // templatized for lockable type
  return common::make_unique<
      common::MultiLockableReference<std::mutex>>(
          mPtr, mSkeletonMutexes.begin(), mSkeletonMutexes.end());

  // templatized for iterator type
  return common::make_unique<
      common::MultiLockableReference<std::set<std::mutex*>::iterator>>(
          mPtr, mSkeletonMutexes.begin(), mSkeletonMutexes.end());
```